drivers: dma_mcux_lpc: Change OTrig sempahore to spinlock

decsny opened the pull request:
Taking semaphore from ISR causes problems, and we also dont want to return an error from dma_config due to otrigs being configured, so replace semaphore with spin lock

Fixes #58457

decsny commented:
> If this really is a shared block for the DMA, and needs a lock, and needs to be usable from an ISR I'd recommend using a spin lock rather than a non blocking semaphore.

I think I understand why you are suggesting that enhancement to this code (it would allow the thread to configure the dma without needing to error out if another thread is messing with otrigs), but just two questions

1. Is it considered a bug fix in zephyr release cycle (would like to get a fix for this issue merged before the release)
2. Is `dma_config` allowed to block a thread's execution like a spinlock would? (The answer to this being a bit unclear is I think what led to this bug in the first place)

teburd commented:
> > If this really is a shared block for the DMA, and needs a lock, and needs to be usable from an ISR I'd recommend using a spin lock rather than a non blocking semaphore.
> 
> I think I understand why you are suggesting that enhancement to this code (it would allow the thread to configure the dma without needing to error out if another thread is messing with otrigs), but just two questions
> 
> 1. Is changing to spinlock considered a bug fix in zephyr release cycle (would like to get a fix for this issue merged before the release)

Replacing a broken semaphore with a working spin lock would be a bug fix in my mind.

> 2. Is `dma_config` allowed to block a thread's execution like a spinlock would? (The answer to this being a bit unclear is I think what led to this bug in the first place)

A spin lock on a single core system is equivalent to a critical section. Is dma_config a time consuming operation? I'd guess not. If not then this should be fine.

Commit:
drivers: dma_mcux_lpc: Replace sem with spinlock

Replace the otrig configuration semaphore with a spinlock.

This will allow the dma_config function to be called from an ISR
and it will no longer shift the burden of waiting to be able to configure
the dma otrigs to the caller of the function, since the driver
will just spin on the lock until it can configure them, instead of
returning an error.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>
